src/bitcoin/Script.mo

```diff
@@ -0,0 +1,89 @@
+import Buffer "mo:base/Buffer";
+import Nat8 "mo:base/Nat8";
+import Array "mo:base/Array";
+import Common "../Common";
+import PublicKey "../ecdsa/Publickey";
+
+module Script {
+  // Supported set of opcodes.
+  public type Opcode = {
+    #OP_DUP;
+    #OP_HASH160;
+    #OP_EQUALVERIFY;
+    #OP_CHECKSIG;
+  };
+
+  // An instruction is either an opcode or data.
+  public type Instruction = {
+    #opcode : Opcode;
+    #data : [Nat8];
+  };
+
+  // A script is an array of instructions.
+  public type Script = [Instruction];
+
+  // Create P2PKH script for the given public key.
+  public func makeP2PKHScript(pk: PublicKey.PublicKey) : Script {
+    return [
+      #opcode(#OP_DUP),
+      #opcode(#OP_HASH160),
+      #data(pk.hash()),
+      #opcode(#OP_EQUALVERIFY),
+      #opcode(#OP_CHECKSIG)
+    ];
+  };
+
+  // Convert given opcode to its byte representation.
+  func encodeOpcode(opcode : Opcode) : Nat8 {
+    return switch (opcode) {
+      case (#OP_DUP) {
+        0x76;
+      };
+      case (#OP_EQUALVERIFY) {
+        0x88;
+      };
+      case (#OP_HASH160) {
+        0xa9;
+      };
+      case (#OP_CHECKSIG) {
+        0xac;
+      };
+    };
+  };
+
+  // Serialize given script to bytes.
+  public func toBytes(script : Script) : [Nat8] {
+    let buf : Buffer.Buffer<Nat8> = Buffer.Buffer<Nat8>(script.size());
+
+    for (instruction in script.vals()) {
+      switch (instruction) {
+        case (#opcode(opcode)) {
+          buf.add(encodeOpcode(opcode));
+        };
+        case (#data data) {
+          // Max data size currently supported.
```

Comment:
So, the maximum size is  (0x4a) 74 bytes?
Where does this bound come from?

Reply:
Larger sizes require handling `PUSHDATA` opcodes which I wasn't planning to do as it's not necessary for p2pkh case. Eventually I added them anyways as they're needed by various test data. So the limit is now removed.